### PR TITLE
gh70 - integrate decodeParams with singleton

### PR DESCRIPTION
**Pull request by zghp**
Screenshot 2024-07-09 at 15 38 26

### CODE REVIEW

File: tests/test_decodeParams.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    args, remaining_args = parser.parse_known_args()
+
+    if args.config:
+        print(f"Using config file: {args.config}")
+        # Load and set the configuration in the singleton
+        singleton.set_config(args.config)
+        # Load and process the config file if needed
+
+    # Run unit tests
+    unittest.main(argv=[sys.argv[0]] + remaining_args, exit=False)
```
Comment (TB-1993) [inline tests/test_decodeParams.py:34]: Shouldn't need to manipulate argv for main to run. We need that to be handled in the singleton or in a RAFTTestMain function.

**Review by TB-1993 — CHANGES_REQUESTED** [review]
Don't forget to clean up all the commented out code too

File: framework/core/decodeParams.py [hunk 2]
```diff
@@ -109,7 +109,7 @@ def decodeDeviceConfig( self ):
         """Decodes the device configuration.
         """
         self.deviceConfig = None
-        deviceConfigFile = self.args.deviceConfigFile
+        deviceConfigFile = self.args.configFile
```
[the remark this answers is not in the document]
Comment (zghp) [reply framework/core/decodeParams.py:112]: tried running exceptionTests.py w/ self.args.deviceConfigFile => no config file 
(but works with self.args.configFile)

File: framework/core/singleton.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        self.config = config
+
+    def get_config(self):
+        return self.config
+    
+    # def get_config(self):
+    #     return self._instance.config
+
+
+singleton = Singleton(log=None)
```
[the remark this answers is not in the document]
Comment (zghp) [reply framework/core/singleton.py:55]: done

File: tests/test_decodeParams.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+def main():
+    parser = argparse.ArgumentParser(description="Run RAFT unit tests.")
+    parser.add_argument('--config', type=str, help='Path to the configuration file')
+    args, remaining_args = parser.parse_known_args()
+
+    if args.config:
+        print(f"Using config file: {args.config}")
+        # Load and set the configuration in the singleton
+        singleton.set_config(args.config)
+        # Load and process the config file if needed
```
[the remark this answers is not in the document]
Comment (zghp) [reply tests/test_decodeParams.py:31]: done

File: framework/core/raftUnittest.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        args, remaining_args = parser.parse_known_args()
+
+        if args.config:
+            print(f"Using config file: {args.config}")
+            # Load and set the configuration in the singleton
+            singleton_instance = Singleton(args.config)
+            singleton_instance.config = decodeParams(args.config)
+
+        # Run unit tests 
+        unittest.main(argv=[sys.argv[0]] + remaining_args, exit=False)
```
Comment (TB-1993) [inline framework/core/raftUnittest.py:31]: Wasn't expecting this function to exist at all.
All argument parsing must be done in decodeParams as part of the Singleton.

File: framework/core/raftUnittest.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+
+class RAFTUnitTestMain(unittest.TestCase):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.deviceManager = SINGLETON.deviceManager
```
Comment (TB-1993) [inline framework/core/raftUnittest.py:38]: Didn't we decide we didn't need this.

At mosts we'd want:
```python
def RAFTmain():
    if SINGLETON.loop:
        for i in SINGLETON.loop:
            unittest.main()
```
But I'd hope there was a better way to do that.

File: tests/test_decodeParams.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+        if params.slotName:
+            print(params.slotName)
+        else:
+            print("No slot name provided")
+
+        if params.loop:
+            print(params.loop)
+        else:
+            print("No loop provided")
```
Comment (TB-1993) [inline tests/test_decodeParams.py:37]: Make this a proper unittest test using asserts. e.g.
```python
def test_decodeParams(self):
    self.assertEqual(SINGLETON.params.args.slot, 1, 'Test the slot value is 1')
```
So the test will pass when I run it with `--slot 1` but fail when I run it with `--slot 2`.

**Review by TB-1993 — CHANGES_REQUESTED** [review]

File: framework/core/raftUnittest.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+            print(f"Using config file: {args.config}")
+            # Load and set the configuration in the singleton
+            singleton_instance = Singleton(args.config)
+            singleton_instance.config = decodeParams(args.config)
+
+        # Run unit tests 
+        unittest.main(argv=[sys.argv[0]] + remaining_args, exit=False)
+
+
+class RAFTUnitTestMain(unittest.TestCase):
```
Comment (Ulrond) [inline framework/core/raftUnittest.py:34]: don't see a test for the testSuite, only a testcase.

File: tests/test_decodeParams.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        expected_rack_name = "rack1"
+        expected_slot_name = "slot1"
+        expected_loop = "1"
+
+        # Assert that the actual values match the expected values
+        self.assertEqual(params.rackName, expected_rack_name, f"rackName does not match the expected value. Got: {params.rackName}")
+        if params.rackName:
+            print(f"rackName: {params.rackName}")
+        else:
+            self.fail("No rack name provided")
```
Comment (TB-1993) [inline tests/test_decodeParams.py:51]: Doesn't the assert fail anyway if params.rackName isn't set?

File: tests/test_decodeParams.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        - The slotName parameter matches the expected value.
+        - The loop parameter matches the expected value.
+        
+
+        Running the test with the following command should pass:
+        $ python tests/test_decodeParams.py --config examples/configs/getting_started_rack_config.yml --rack rack1 --slotName slot1 --loop 1
+        
+        Running the test with any of the following commands should fail:
+        $ python tests/test_decodeParams.py --config examples/configs/getting_started_rack_config.yml
+        $ python tests/test_decodeParams.py --config examples/configs/getting_started_rack_config.yml --rack rack2 --slotName slot3 --loop 1
```
Comment (TB-1993) [inline tests/test_decodeParams.py:36]: Can we do a run with every possible switch and then check all of them in this test?

**Review by TB-1993 — APPROVED** [review]

**Comment by TB-1993** [comment]
Ship it!

**Pull request merged**